luatest/server_v2.lua

```diff
@@ -0,0 +1,503 @@
+    end
+    io.flush() -- attempt to flush stdout == log fd
+    local filesize = file:seek(0, 'SEEK_END')
+    if filesize == nil then
+        fail('Failed to get log file size')
+    end
+    local bytes = bytes_num or 65536 -- don't read whole log - it can be huge
+    bytes = bytes > filesize and filesize or bytes
+    if file:seek(-bytes, 'SEEK_END') == nil then
+        fail('Failed to seek log file')
+    end
+    local found, buf
+    repeat -- read file in chunks
+        local s = file:read(2048)
+        if s == nil then
+            fail('Failed to read log file')
+        end
+        local pos = 1
+        repeat -- split read string in lines
+            local endpos = string.find(s, '\n', pos)
+            endpos = endpos and endpos - 1 -- strip terminating \n
+            local line = string.sub(s, pos, endpos)
+            if endpos == nil and s ~= '' then
+                -- Line doesn't end with \n or EOF, append it to buffer
+                -- to be checked on next iteration.
+                buf = buf or {}
+                table.insert(buf, line)
+            else
+                if buf ~= nil then
+                    -- Prepend line with buffered data.
+                    table.insert(buf, line)
+                    line = table.concat(buf)
+                    buf = nil
+                end
+                if string.match(line, 'Tarantool %d.%d+.%d+') and reset then
+                    found = nil -- server was restarted, reset search
+                else
+                    found = string.match(line, pattern) or found
+                end
+            end
+            pos = endpos and endpos + 2 -- jump to char after \n
+        until pos == nil
+    until s == ''
+    file:close()
+    return found
+end
+
+--
+-- Replication
+--
+
+--- Assert that the server follows the source node with the given ID.
+-- Meaning that it replicates from the remote node normally, and has already
+-- joined and subscribed.
+--
+-- @number server_id Server ID.
+function Server:assert_follows_upstream(server_id)
+    local status = self:exec(function(id)
+        return box.info.replication[id].upstream.status
+    end, {server_id})
+    local msg = ('%s: server does not follow upstream'):format(self.alias)
+    assertions.assert_equals(status, 'follow', msg)
+end
+
+-- Election
+
+--- Get the election term as seen by the server.
+--
+-- @return number
+function Server:get_election_term()
+    return self:exec(function() return box.info.election.term end)
+end
+
+--- Wait for the server to reach at least the given election term.
+--
+-- @string term Election term to wait for.
+function Server:wait_for_election_term(term)
+    return wait_for_condition('election term', self, self.exec, self, function(t)
+        return box.info.election.term >= t
+    end, {term})
+end
+
+--- Wait for the server to enter the given election state.
+-- Note that if it becomes a leader, it does not mean it is already writable.
```

Comment:
It could be useful to see «use method `foo` to check that it is writable»

Reply:
Not sure about this.